### ISSUE TITLE
feat: detect, show and use top-level directory of cmake

### ISSUE THREAD

**Issue by matu3ba** [cdelledonne/vim-cmake#28]
__Reasoning__
It is quite annoying to be forced to always jump to the top level directory, but we also dont want to mess up the build. For example, git submodules or external projects on wrong invocation of cmake may lead to this.
(the CMakeList.txt does not hint that the code is used as submodule)

__Status Line__
I It would be good to show the top-level directory in the statusline, but probably sufficient to show the top-level directory folder name in the status line. Adding instructions for doing this to the README would be great.

__HOW__
We can check `git rev-parse --show-toplevel` and `git rev-parse --show-superproject-working-tree`. Does cmake offer a similar solution? I did not find any yet.

**Comment by cdelledonne** [comment]
Hi, thank you for posting this.

I agree that the detection of the top-level directory could use some improvement. Other than what you suggest, the top-level directory should also be computed each time it is needed, and not only when the plugin is loaded (this is the current behaviour), so that if the user changes the CWD, the path to the top-level directory is updated accordingly.

At the moment, I am quite busy with some other stuff and cannot work on this plugin too much, would you be interested in preparing a pull request for this yourself? Otherwise I can take this up in a few weeks.

To recap, what we would need is the following

- By default, detect top-level directory using `git rev-parse`
- If that fails for some reason—e.g., we're not in a Git directory, fallback to using the current method that searches for root markers defined in `g:cmake_root_markers`
- Recompute top-level directory each time it is needed
- Support showing the top-level directory in the statusline
- Perhaps add a config option to disable showing the top-level directory in the statusline

Am I missing something?

**Comment by matu3ba** [comment]
You are missing the case, if we are in a git submodule folder. Aside, that looks good to me.
I can not really tell you, when I find the time to do this as I never have written alot vimscript before.

**Comment by cdelledonne** [comment]
> You are missing the case, if we are in a git submodule folder.

Doesn't that still fall under "By default, detect top-level directory using `git rev-parse`"? The two options you listed, `--show-toplevel` and `--show-superproject-working-tree`, should give us what we need, right?

>  I can not really tell you, when I find the time to do this as I never have written alot vimscript before.

Vimscript is a little ugly but not too difficult. If you are interested in submitting a PR, whenever you find time, I can review your work step by step.
